pymaster/field.py

```diff
@@ -658,3 +681,176 @@ def get_ell_sampling(self):
+            radians are provided instead.
+    """
+    def __init__(self, positions, weights, field, lmax, lmax_mask=None,
+                 spin=None, positions_rand=None, weights_rand=None, beam=None,
+                 field_is_weighted=False, lonlat=False):
+        # 0. Preliminary initializations
+        if ut.HAVE_DUCC:
+            self.sht_calculator = 'ducc'
+        else:
+            raise ValueError("DUCC is needed, but currently not installed.")
+
+        # These first attributes are compulsory for all fields
+        self.lite = True
+        self.mask = None
+        self.beam = None
+        self.n_iter = None
+        self.n_iter_mask = None
+        self.pure_e = False
+        self.pure_b = False
+        self._Nw = 0.
+        self._Nf = 0.
+        self._alpha = None
+        self.ainfo = ut.NmtAlmInfo(lmax)
+
+        # The remaining attributes are only required for non-lite maps
+        self.maps = None
+        self.temp = None
+        self.alm_temp = None
+        self.minfo = None
+        self.n_temp = 0
+        do_clustering = positions_rand is not None and weights_rand is not None
+
+        # Can only have field or do clustering
+        if field is not None and do_clustering:
+            raise ValueError("If positions_rand and weights_rand are "
+                             "provided, field must be None.")
+
+        # Sanity checks on positions and weights
+        positions = np.array(positions, dtype=np.float64)
+        weights = np.array(weights, dtype=np.float64)
+        pw_list = [(positions, weights)]
+        if do_clustering:
+            positions_rand = np.array(positions_rand, dtype=np.float64)
+            weights_rand = np.array(weights_rand, dtype=np.float64)
+            pw_list += [(positions_rand, weights_rand)]
+        for p, w in pw_list:
+            if np.shape(p) != (2, len(w)):
+                raise ValueError("Positions must be 2D array of shape"
+                                 " (2, len(weights)).")
+            if lonlat:
+                p[0] = np.radians(90. - p[1])
+                p[1] = np.radians(p[0])
+            if not (np.logical_and(p[0] >= 0., p[0] <= np.pi)).all():
+                if lonlat:
+                    raise ValueError(
+                        "Second dimension of positions must be latitude in "
+                        "degrees, between -90 and 90."
+                    )
+                else:
+                    raise ValueError(
+                        "First dimension of positions must be colatitude in "
+                        "radians, between 0 and pi."
+                    )
+
+        # Compute mask shot noise
+        self._Nw = np.sum(weights**2.)/(4.*np.pi)
+
+        # 1. Compute mask alms and beam
+        # Sanity checks
+        if lmax_mask is None:
+            lmax_mask = 4*lmax
+        self.ainfo_mask = ut.NmtAlmInfo(lmax_mask)
+        # Mask alms
+        lmax_alm = lmax if do_clustering else lmax_mask
```

Comment:
Why treat `do_clustering` differently?

Reply:
Because in the clustering case, we do `alm = (self.alm_mask - self._alpha*alm_rand).reshape(1, -1)` in l. 835 below, so `alm_mask` and `alm_rand` (which is expected to have `lmax=lmax`) must have the same shape.

Reply:
yes, good point. I realised afterwards.